File: first_non_repeating_character/first_non_repeating_character.java
import java.util.*;

public class FirstNonRepeatingCharacter {
    public static int firstNonRepeatingCharacter(String str) {
        Map<Character, Boolean> characters = new HashMap<>();
        for (char c : str.toCharArray()) {
            characters.put(c, !characters.containsKey(c));
        }
        for (int i = 0; i < str.length(); i++) {
            if (characters.get(str.charAt(i))) {
                return i;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println(firstNonRepeatingCharacter("abcdcaf"));
    }
}

File: first_non_repeating_character/first_non_repeating_character.ts
function firstNonRepeatingCharacter(str: string): number {
  const characters: { [key: string]: boolean } = {};
  for (const c of str) {
      characters[c] = characters[c] === undefined;
  }
  for (let i = 0; i < str.length; i++) {
    if (characters[str[i]]) {
      return i;
    }
  }
  return -1;
}

console.log(firstNonRepeatingCharacter("abcdcaf"));
